# CLS3 Final data release
## GTF tags description
GTF files generated from the CLS3 data:
    <li><a href="#CLS-anchored-models">CLS anchored models</a></li>
    <li><a href="#CLS-transcripts">CLS transcripts</a></li>
    <li><a href="#CLS-loci">CLS loci</a>
    <li><a href="#Target-files">Target files</a>
    
## CLS anchored models
The CLS3 long-read data is processed through LyRic to obtaing high-confidence transcript models per sample through starting from pacBioSII- and ONT-sequenced long reads. Further, to collect a comprehensive set of transcripts
from the different tissues and technologies, the models were "anchored" according to their support at 5' and 3' ends ([anchorTranscriptsEnds.pl](https://github.com/guigolab/LyRic/blob/master/utils/anchorTranscriptsEnds.pl)), and merged together ([tmerge](https://github.com/guigolab/tmerge)). 

Each feature in the table is associated to a _gene_id_ and _transcript_id_ attributes, specifying the unique identifier as generated by LyRic. The `transcript` features are additionally endowed with the following attributes:
||||
|-|-|-|
1 | target | comma-separated list of the genomic regions targeted by the pipeline. For each target we report, in order, the *source database*, the *identifier*, the *chromosome*, *start* and *end* coordinates, and the *strand*.
2 | endSupport | a value among **polyAOnlySupported**, **cageOnlySupported**, **cagePolyASupported**, **noCageNoPolyASupported**, indicating the type of support available for the transcript model.
3 | spliced | either **spliced** or **unspliced**, indicating whether the associated transcript model is composed of different exons or remains unspliced.
4 | refCompare | result of gffcompare against Gencode annotation *v27* for human and *vM16* for mouse. The original codes have been further collapsed to obtain the following categories: **Antisense** (corresponding gffcompare codes s, x), **Equal** (=), **Extends** (k), **Included** (c), **Intergenic** (u), **Intronic** (i), **revIntronic** (y), **Overlaps** (j, e, m, o, n), **runOn** (p)
5 | currentCompare | result of gffcompare against Gencode annotation *v43* for human and *vM31* for mouse, same categories as above.Â 
6 | sampleN | integer value indicating the number of transcripts across all samples merged into the respective transcript model
7 | samplesMetadata | a list of mnemotechnics samples IDs the collapsed transcripts were encountered in. See [Samples Metadata](#samples-metadata)
8 | expression | decimal value corresponding to the expression level of the transcript, expressed as RPM. The order of the values matches the order of the samples the transcript belongs to in the previous tag.
9 | artifact | list of tags reflecting whether the model is deemed artefactual and why. See [Artefacts Tags](#artefacts-tags)

### Samples Metadata
The samples IDs have been generated in a way to keep track of as many metadata as possible. The names are composed as follows:
||||
|-|-|-|
0 | Fixed prefix | SID
1 | Single letter code for the organism | `H` (human), `M` (mouse)
2 | Two letter code indicating the tissue | See [Tissue codes](#tissue-codes)
3 | Single letter code for the stage | `A` (adult), `E` (embryo), `P` (placenta)
4 | Single letter code for the sequencing technology | `O` (ONT), `P` (PacBio)
5 | Single letter code for capture status | `P` (pre-capture), `C` (post-capture)
6 | Two digit code for the biological replicate | - 
7 | Two digit code for the technical replicate | - 

### Artefacts Tags
||||
|-|-|-|
0 | no | Genuine model.
1 | oppStrandMismap | i.e., Opposite strand mismapping. Models mapping on opposite strand of annotated coding loci.
2 | polyASJdisag | i.e., PolyA - Splice Junction disagreement. Highlights possible problems during mapping steps, where polyA and splice junction provide contradictory information over the read strand.
3 | pseudogeneOverlap | Models contained within an annotated pseudogene locus. This are supposedly generated from parent gene but wrongly assigned due to the polyA stretch at 3'.
4 | recountSlt50 | Models in which any of the splice junctions doesn't meet the minimum threshold of recount support (50).
5 | spliceSiteMisalign | i.e., Splice site misalignments. Highlights uncertainty in splice junction placement at mapping.
6 | tRepeatOverlap | i.e., Tandem repeat overlapping.

### Tissue Codes
|||
|-|-|
Brain | `Br`
Heart | `He`
Liver | `Li`
WBCs | `Wb`
ESC | `Wb`
iPSC | `Wb`
Testis | `Te`
Placenta | `Pl`
Tpool | `Tp`
Cpool | `Cp`

Some examples of aliases and their meaning
 * `SIDMBrEPP0101`: mouse brain embryo PacBio pre-capture biologicalReplicate01 technicalReplicate01
 * `SIDHWbAOC0103`: human whiteBlood adult ont post-capture biologicalReplicate01 technicalReplicate03
 * `SIDHPlPPP0202`: human placenta placenta PacBio pre-capture biologicalReplicate02 technicalReplicate02

## CLS transcripts
To further reduce the redundancy of the aforementioned transcripts collection, models sharing the same structure were merged together, ignoring eventual variation at the terminal exons, and therefore neglecting end-support information. 
In addition to the attributes specified before, _contained_anchTMs_ has been added to keep track of which anchored transcripts were merged within each chain. 

While all the attributes have been recomputed to address the new transcripts structure (e.g., _refCompare_ and _currentCompare_), the _endSupport_ tag specifies the highest support 
available among the anchored transcripts collapsed.


## CLS loci
With the intent of grouping together different models in uniquely identifiable loci, we clustered CLS transcripts into regions of continuous transcription.
Eventually, transcripts sharing any overlap on the same strand have been brought together into a single locus, preserving their structure. 

Each exon has the transcript level attributes obtained through tmerge, description [here](https://github.com/guigolab/tmerge?tab=readme-ov-file#output). Each locus can be linked to the underlying set of anchored transcripts by means of _contains_ attribute. The following additional attributes
have been introduced to specifically track the eventual exonic overlap of the defined transcript with a previously annotated entry; _overlapping_gencode_gene_, _overlapping_gencode_transcript_.
Finally, _gLlevel_gencodeGeneOverlaps_ specifies the locus overlap with gencode genes at the exonic level. 
