███████╗██╗ ██████╗ ██╗   ██╗██████╗ ███████╗    ██████╗  █████╗      █████╗ ███╗   ██╗██████╗     ███████╗ ██╗███████╗██████╗ 
██╔════╝██║██╔════╝ ██║   ██║██╔══██╗██╔════╝    ╚════██╗██╔══██╗    ██╔══██╗████╗  ██║██╔══██╗    ██╔════╝███║╚════██║██╔══██╗
█████╗  ██║██║  ███╗██║   ██║██████╔╝█████╗       █████╔╝███████║    ███████║██╔██╗ ██║██║  ██║    ███████╗╚██║    ██╔╝██████╔╝
██╔══╝  ██║██║   ██║██║   ██║██╔══██╗██╔══╝       ╚═══██╗██╔══██║    ██╔══██║██║╚██╗██║██║  ██║    ╚════██║ ██║   ██╔╝ ██╔══██╗
██║     ██║╚██████╔╝╚██████╔╝██║  ██║███████╗    ██████╔╝██║  ██║    ██║  ██║██║ ╚████║██████╔╝    ███████║ ██║   ██║  ██████╔╝
╚═╝     ╚═╝ ╚═════╝  ╚═════╝ ╚═╝  ╚═╝╚══════╝    ╚═════╝ ╚═╝  ╚═╝    ╚═╝  ╚═╝╚═╝  ╚═══╝╚═════╝     ╚══════╝ ╚═╝   ╚═╝  ╚═════╝ 
                                                                                                                               
# Load R libraries
```{r}
if (!require("pacman")) install.packages("pacman")
pacman::p_load(ggplot2,
               scales,
               dplyr)
```

# Prepare environment
```{bash echo=FALSE, engine.opts='-l'}
# Remove old results
rm -r ../Data/.temp/Figure_3A+S17B/

# Create directory tree
mkdir ../Data/.temp/Figure_3A+S17B/

# Move to working dir
cd ../Data/.temp/Figure_3A+S17B/

# Add chess and lncBook
## Catalogues <- exclude chrM, keep unspliced
zcat ../../Source/Catalogs/noncode.gtf.gz | ../../../Utils/gff2bed_full.pl - | awk 'BEGIN{FS=OFS="\t"}$6!="."' | awk '$1 ~ /^chr[0-9XY]{1,2}$/ {print $0}' | grep "\S" | ../../../Utils/bed12togff - > noncode.hg38.gtf
zcat ../../Source/Catalogs/refseq.gtf.gz | ../../../Utils/gff2bed_full.pl - | awk 'BEGIN{FS=OFS="\t"}$6!="."' | awk '$1 ~ /^chr[0-9XY]{1,2}$/ {print $0}' | grep "\S" | ../../../Utils/bed12togff - > refseq.hg38.gtf
zcat ../../Source/Catalogs/mitrans.gtf.gz | ../../../Utils/gff2bed_full.pl - | awk 'BEGIN{FS=OFS="\t"}$6!="."' | awk '$1 ~ /^chr[0-9XY]{1,2}$/ {print $0}' | grep "\S" | ../../../Utils/bed12togff - > mitrans.hg38.gtf
zcat ../../Source/Catalogs/gencode20+.gtf.gz | ../../../Utils/gff2bed_full.pl - | awk 'BEGIN{FS=OFS="\t"}$6!="."' | awk '$1 ~ /^chr[0-9XY]{1,2}$/ {print $0}' | grep "\S" | ../../../Utils/bed12togff - > gencode20+.hg38.gtf
zcat ../../Source/Catalogs/bigtrans.gtf.gz | ../../../Utils/gff2bed_full.pl - | awk 'BEGIN{FS=OFS="\t"}$6!="."' | awk '$1 ~ /^chr[0-9XY]{1,2}$/ {print $0}' | grep "\S" | ../../../Utils/bed12togff - > bigtrans.hg38.gtf
zcat ../../Source/Catalogs/fantomCat.gtf.gz | ../../../Utils/gff2bed_full.pl - | awk 'BEGIN{FS=OFS="\t"}$6!="."' | awk '$1 ~ /^chr[0-9XY]{1,2}$/ {print $0}' | grep "\S" | ../../../Utils/bed12togff - > fantomCat.hg38.gtf
# Add chess and lncBook
zcat ../../Source/Catalogs/chess3_1_2_GRCh38_lncRNA.gtf.gz | ../../../Utils/gff2bed_full.pl - | awk 'BEGIN{FS=OFS="\t"}$6!="."' | awk '$1 ~ /^chr[0-9XY]{1,2}$/ {print $0}' | grep "\S" | ../../../Utils/bed12togff - > chess3_1_2.hg38.gtf
zcat ../../Source/Catalogs/lncRNA_LncBookv2_0_GRCh38.gtf.gz | ../../../Utils/gff2bed_full.pl - | awk 'BEGIN{FS=OFS="\t"}$6!="."' | awk '$1 ~ /^chr[0-9XY]{1,2}$/ {print $0}' | grep "\S" | ../../../Utils/bed12togff - > lncBookv2_0.hg38.gtf

# Add full and nice gencode 27 and 47 
zcat ../../Source/Catalogs/gen47.gtf.gz | awk 'BEGIN{FS=OFS="\t"}$7!="."' | awk '$1 ~ /^chr[0-9XY]{1,2}$/ {print $0}' | awk '$3=="exon"' | grep "\S" > gen47.hg38.gtf
zcat ../../Source/Catalogs/gen27.gtf.gz | awk 'BEGIN{FS=OFS="\t"}$7!="."' | awk '$1 ~ /^chr[0-9XY]{1,2}$/ {print $0}' | awk '$3=="exon"' | grep "\S" > gen27.hg38.gtf

# Include concatenated catalogs
## Concatenate human catalogs into 1 file (ENST00000610119.1 has inconsistent strand); remove chrM
cat noncode.hg38.gtf refseq.hg38.gtf mitrans.hg38.gtf gencode20+.hg38.gtf bigtrans.hg38.gtf fantomCat.hg38.gtf | sort | uniq | grep -vFw "ENST00000610119.1" | grep "\S" | ../../../Utils/sortgff - | ../../../Utils/tmerge --exonOverhangTolerance 8 --tmPrefix IC - | grep '\S' > H_concat_catalogs.hg38.gtf

# Prepare config file (necessary for anchored merging (heatmap, boxplot) and barplots)
## Ends
echo "5" >> ends.config
echo "3" >> ends.config

# Prepare Config files
for catalogue in `ls *.hg38.gtf`; do echo $catalogue >> catalogs.config; done
```



# Generate data for Gold (CapTrap-CLS_lncRNAs) UnifiedCatalogue >> heatmap, boxplot and barplot <<
```{bash echo=FALSE, engine.opts='-l'}
# Move to working dir
cd ../Data/.temp/Figure_3A+S17B/

# Remove files from previous run
rm ../../../Figure_3A+S17B/Figure_3A+S17B_heatmap_results.tsv

# Handle errors
set -e          # exit on any non-0 exit status
set -o pipefail # exit on any non-0 exit status in pipe
#set -m

# Create LOG file
LOG_LOCATION=./LOGFILE.log
exec > >(tee -i $LOG_LOCATION)
exec 2>&1

echo "Log saved to: [ $LOG_LOCATION ]"

# Run buildLoci for all except gencode 27 and gencode 47
echo "Running buildLoci"
while read file || [ -n "$file" ]
  do
      name=`echo $file | awk -F'.' '{print $1}'`
      echo "Name: $name"
      if [[ "$name" == "gen27" ]] || [[ "$name" == "gen47" ]]; then
         ../../../Utils/gff2gff.pl $name.hg38.gtf > $name.hg38.gene.gtf
         ../../../Utils/extract_locus_coords.pl $name.hg38.gene.gtf > $name.hg38.gene.bed
      
      else
         bedtools intersect -s -wao -a $file -b $file | ../../../Utils/buildLoci.pl - > $name.gene.tmp.gtf
         ../../../Utils/gff2gff.pl $name.gene.tmp.gtf > $name.hg38.gene.gtf
         ../../../Utils/extract_locus_coords.pl $name.hg38.gene.gtf > $name.hg38.gene.bed
      fi
  done < ./catalogs.config

# Run intersecting
echo "Running intersecting"
while read file_a || [ -n "$file_a" ]
  do
      name_a=`echo $file_a | awk -F'.' '{print $1}'`
      echo "Name_a: $name_a"
        
      while read file_b || [ -n "$file_b" ]
        do
          name_b=`echo $file_b | awk -F'.' '{print $1}'`
          echo "Name_b: $name_b"
          
          bedtools intersect -s -wa -wb -f 1.00 -F 1.00 -e -a $name_a.hg38.gene.bed -b $name_b.hg38.gene.bed > $name_a.intersecting.$name_b.bed
          
          size=`cat $name_a.hg38.gene.bed | sort | uniq | wc -l`
          intersecting=`cat $name_a.intersecting.$name_b.bed | awk '{print $4}' | sort | uniq | wc -l`
          percentage=`echo "scale=2 ; $intersecting / $size" | bc`
          echo -e "$name_a\t$name_b\t$size\t$intersecting\t$percentage" >> ../../../Figure_3A+S17B/Figure_3A+S17B_heatmap_results.tsv
        done < ./catalogs.config
  done < ./catalogs.config
```

# Plot heatmap and boxplot
```{r}
# Clean the env
rm(list = ls())

# Load data
df <- read.table("./Figure_3A+S17B_heatmap_results.tsv", header = F, sep = "\t") %>%
  rename(
    name_a = "V1",
    name_b = "V2",
    size = "V3",
    intersecting = "V4",
    percentage = "V5"
  ) %>%
  filter(name_a != "cls") %>%
  filter(name_b != "cls") %>%
  filter(name_a != "tmerged_anchored_Human_CapTrap-CLS_lncRNAs+gen27" & name_b != "tmerged_anchored_Human_CapTrap-CLS_lncRNAs+gen27") %>%
  filter(name_a != "tmerged_anchored_Human_CapTrap-CLS_lncRNAs+gen47" & name_b != "tmerged_anchored_Human_CapTrap-CLS_lncRNAs+gen47") %>%
  # Add column for boxplot facet
  mutate(category_a = ifelse((name_a == "lncBookv2_0" | name_a == "chess3_1_2"), "Non-targeted lncRNA annotations", NA)) %>%
  mutate(category_a = ifelse((name_a == "gen27" | name_a == "gen47"), "GENCODE", category_a)) %>%
  mutate(category_a = ifelse((is.na(category_a)), "Targeted lncRNA annotations", category_a)) %>%
  mutate(category_b = ifelse((name_b == "lncBookv2_0" | name_b == "chess3_1_2"), "Non-targeted lncRNA annotations", NA)) %>%
  mutate(category_b = ifelse((name_b == "gen27" | name_b == "gen47"), "GENCODE", category_b)) %>%
  mutate(category_b = ifelse((is.na(category_b)), "Targeted lncRNA annotations", category_b))

df


levels_a <- c(
  "lncBookv2_0",
  "chess3_1_2",
  "refseq",
  "noncode",
  "mitrans",
  "gencode20+",
  "fantomCat",
  "bigtrans",
  "H_concat_catalogs",
  "gen47",
  "gen27"
)

levels_b <- c(
  "gen27",
  "gen47",
  "H_concat_catalogs",
  "bigtrans",
  "fantomCat",
  "gencode20+",
  "mitrans",
  "noncode",
  "refseq",
  "chess3_1_2",
  "lncBookv2_0"
)


labels_a <- c(
  "LncBook v2.0",
  "CHESS 3.1.2",
  "RefSeq",
  "NONCODE",
  "MiTranscriptome",
  "GENCODE20+",
  "Fantom Cat",
  "BIGTranscriptome",
  "lncRNA-merge",
  "GENCODE 47",
  "GENCODE 27"
)

labels_b <- c(
  "GENCODE 27",
  "GENCODE 47",
  "lncRNA-merge",
  "BIGTranscriptome",
  "Fantom Cat",
  "GENCODE20+",
  "MiTranscriptome",
  "NONCODE",
  "RefSeq",
  "CHESS 3.1.2",
  "LncBook v2.0"
)

# Create heatmap - Figure S17B
df %>%
  ggplot(data = .) +
  geom_tile(
    data = df %>% filter(percentage == 1.0),
    aes(
      x = factor(name_b, levels = rev(levels_a), labels = rev(labels_a)),
      y = factor(name_a, levels = rev(levels_b), labels = rev(labels_b))
    ), fill = "#014636"
  ) +
  geom_tile(
    data = df %>% filter(percentage < 1.0),
    aes(
      x = factor(name_b, levels = rev(levels_a), labels = rev(labels_a)),
      y = factor(name_a, levels = rev(levels_b), labels = rev(labels_b)),
      fill = percentage
    )
  ) +
  geom_text(
    size = 5,
    aes(
      x = factor(name_b, levels = rev(levels_a), labels = rev(labels_a)),
      y = factor(name_a, levels = rev(levels_b), labels = rev(labels_b)),
      label = (scales::percent(percentage))
    )
  ) +
  geom_text(
    size = 3.2,
    aes(
      x = factor(name_b, levels = rev(levels_a), labels = rev(labels_a)),
      y = factor(name_a, levels = rev(levels_b), labels = rev(labels_b)),
      label = paste0(
        "\n\n",
        "(", scales::comma(intersecting), "/", scales::comma(size), ")"
      )
    )
  ) +
  scale_fill_gradientn(
    colours = c("#FFF7FB", "#ECE2F0", "#D0D1E6", "#A6BDDB", "#67A9CF", "#3690C0", "#02818A", "#016C59", "#014636"),
    values = rescale(c(0, 0.125, 0.250, 0.375, 0.500, 0.625, 0.775, 0.900, 1.0)),
    guide = "colorbar",
    labels = scales::label_percent(),
    breaks = c(0.25, 0.50, 0.75, 1.0),
    n.breaks = 4
  ) +
  theme(
    axis.title.x = element_blank(),
    axis.title.y = element_blank(),
    axis.ticks = element_line(colour = NA),
    plot.title = element_text(hjust = 0.5),
    panel.background = element_rect(fill = NA),
    panel.grid.major = element_line(colour = NA),
    axis.text.x = element_text(
      colour = "black",
      face = "bold", angle = 90, hjust = 1
    ),
    axis.text.y = element_text(
      colour = "black",
      face = "bold"
    )
  ) +
  labs(title = "Figure S17B", fill = "% of genes") -> plot_Figure_S17B

# Create boxplot - Figure 3A
df %>%
  filter(name_a != name_b) %>%
  filter((name_b != "gen27" & name_a == "gen27") | (name_b != "gen27" & name_a == "gen47") | (name_b != "gen47" & name_a == "gen27") | (name_b != "gen47" & name_a == "gen47") | (name_b == "gen27") | (name_b == "gen47")) %>%
  ggplot(data = ., aes(
    x = factor(name_b, levels = levels_b, labels = labels_b),
    y = percentage
  )) +
  geom_boxplot(color = "#0466c8") +
  geom_dotplot(aes(fill = factor(name_a, levels = levels_a, labels = labels_a)),
    binaxis = "y", stackdir = "center", dotsize = 1
  ) +
  theme_bw() +
  scale_fill_manual(values = c(
    "#f7e49a", "#f8a656", "#fece79", "#f48170",
    "#e4b7d5", "#8b8bc3", "#94cae3", "#da7422",
    "#b2967d", "#25a18e", "#00a5cf"
  )) +
  ggtitle("Figure 3A") +
  ylab("percentage") +
  scale_y_continuous(labels = percent, lim = c(0, 1)) +
  theme(legend.position = "top") +
  theme(
    axis.text.x = element_text(size = 12, angle = 60, hjust = 1, colour = "black"),
    axis.ticks.x = element_blank(),
    axis.text.y = element_text(size = 12, colour = "black"),
    legend.text = element_text(size = 12),
    plot.title = element_text(size = 15, hjust = 0.5),
    legend.title = element_blank(),
    axis.title.x = element_blank(),
    axis.title.y = element_text(size = 14),
    strip.text.x = element_text(size = 18),
    strip.text.y = element_text(size = 18)
  ) +
  facet_grid(. ~ factor(category_b, levels = c("GENCODE", "Targeted lncRNA annotations", "Non-targeted lncRNA annotations")), space = "free", scales = "free", labeller = label_wrap_gen(18)) -> plot_Figure_3A

# Diplay plots
plot_Figure_S17B

plot_Figure_3A
```

## Save as pdf <- heatmap (Figure S17B)
```{r}
pdf(file = "./plot_Figure_S17B.pdf", width = 14, height = 9)
plot_Figure_S17B
```

## Save as pdf <- boxplot (Figure 3A)
```{r}
pdf(file = "./plot_Figure_3A.pdf", width = 14, height = 9)
plot_Figure_3A
```
