░██████████░██                                               ░██████     ░██     ░██          ░██ 
░██                                                         ░██   ░██  ░████   ░████          ░██ 
░██        ░██ ░████████ ░██    ░██ ░██░████  ░███████     ░██           ░██     ░██    ░████████ 
░█████████ ░██░██    ░██ ░██    ░██ ░███     ░██    ░██     ░████████    ░██     ░██   ░██    ░██ 
░██        ░██░██    ░██ ░██    ░██ ░██      ░█████████            ░██   ░██     ░██   ░██    ░██ 
░██        ░██░██   ░███ ░██   ░███ ░██      ░██            ░██   ░██    ░██     ░██   ░██   ░███ 
░██        ░██ ░█████░██  ░█████░██ ░██       ░███████       ░██████   ░██████ ░██████  ░█████░██ 
                     ░██                                                                          
               ░███████                                                                           

# Prepare environment                                                                 
```{bash echo=FALSE, engine.opts='-l'}
# +++++++++++++++++++++++++++++++++++++++++++++
# REQUIREMENTS
# +++++++++++++++++++++++++++++++++++++++++++++
# 1) Catalogs in gtf.gz file format in ./Data/M_catalogs/ directory.
# 2) Additional files in ./Data/Additional/ directory.
# 3) Installed bedtools v2.31.0
# 4) Utils in ../../Utils/ directory:
#     - bed12togff
#     - buildLoci.pl
#     - gffToHash.pm
#     - hashToGff.pm
#     - extractTranscriptEndsFromBed12.pl
#     - gff2bed_full.pl
#     - join.py
#     - sortbed

# +++++++++++++++++++++++++++++++++++++++++++++
# DESCRIPTION
# +++++++++++++++++++++++++++++++++++++++++++++
# Takes data in gtf format from ./Data/M_catalogs/ and calculates Figure S11d plot input


##########################################################################################
# Clean-up generated data
##########################################################################################
echo '++++++ Cleaning up data from previous run ++++++'
rm -r .temp_s11d/
rm figure_s11d_plot_input.tsv
rm figure_s11d_recount_input.tsv
echo '>> Done <<'

##########################################################################################
# Create directory structure
##########################################################################################
echo '++++++ Creating necessary directory structure ++++++'
# Dir for .temp data
mkdir -p .temp_s11d/
mkdir .temp_s11d/M_catalogs/
mkdir .temp_s11d/Cage/
mkdir .temp_s11d/PolyA/
mkdir .temp_s11d/Fl/
mkdir .temp_s11d/Decompressed/
mkdir .temp_s11d/Recount_support/
echo '>> Done <<'

# Handle errors
set -e          # exit on any non-0 exit status
set -o pipefail # exit on any non-0 exit status in pipe

# Go to working directory

cd .temp_s11d/

# Copy M_catalogs
cp ../../Data/M_catalogs/gencodeM3+.gtf.gz ./M_catalogs/
cp ../../Data/M_catalogs/genM16.gtf.gz ./M_catalogs/
cp ../../Data/M_catalogs/genM36_mm10.gtf.gz ./M_catalogs/genM36.gtf.gz
cp ../../Data/M_catalogs/refSeq.gtf.gz ./M_catalogs/
cp ../../Data/M_catalogs/NONCODE.gtf.gz ./M_catalogs/

##########################################################################################
##########################################################################################
# Generate data for FIGURE s11d
##########################################################################################
##########################################################################################

##########################################################################################
# % Support
##########################################################################################

echo 'Get fl-tx'

for file in `ls ./M_catalogs/*gtf.gz | awk -F'/' '{print $NF}' | sed 's/.gtf.gz//g'`
    do
        echo $file
        zcat ./M_catalogs/$file.gtf.gz | ../../Utils/gff2bed_full.pl - | awk 'BEGIN{FS=OFS="\t"}$6!="."' | awk '$1 ~ /^chr[0-9XYM]{1,2}$/ {print $0}' | awk '$10 > 1 {print $0}' | gzip > $file.bed12.gz

        for end in 3 5
        do
            echo $end
            zcat $file.bed12.gz | ../../Utils/extractTranscriptEndsFromBed12.pl $end | ../../Utils/sortbed | gzip > $file.$end.bed.gz
        done

        ## polyA
        echo '++++++ Looking for polyA supported TM ++++++'
        for end in 3 5
        do
            zcat $file.$end.bed.gz | ../../Utils/sortbed | bedtools slop -s -l 50 -r 10 -i stdin -g ../../Data/Additional/mm10.chrom.sizes | bedtools intersect -u -s -a stdin -b ../../Data/Additional/mm10.polyAsignals.bed | gzip > $file.$end.bed.vspolyAsignals.bedtsv.gz
        done
        echo '>> Done <<'

        ## CAGE FANTOM phase 1+2
        echo '++++++ Looking for CAGE supported TM ++++++'
        for end in 3 5
        do
            zcat $file.$end.bed.gz | ../../Utils/sortbed | bedtools slop -s -l 50 -r 50 -i stdin -g ../../Data/Additional/mm10.chrom.sizes | bedtools intersect -u -s -a stdin -b ../../Data/Additional/mm10_fair+new_CAGE_peaks_phase1and2.bed | gzip > $file.$end.bed.vsCage.fantom.bedtsv.gz
        done
        echo '>> Done <<'

        ## Extract Cage and polyA supported transcripts
        echo '++++++ Extracting CAGE and polyA supported transcripts ++++++'
        zcat $file.5.bed.vsCage.fantom.bedtsv.gz | cut -f4 | sed 's/,/\n/g' | sort | uniq > $file.tx.cage
        zcat $file.3.bed.vspolyAsignals.bedtsv.gz | cut -f4 | sed 's/,/\n/g' | sort | uniq > $file.tx.polyA
        echo '>> Done <<'

        ## Get fl-tx for Unified
        ../../Utils/join.py -a $file.tx.cage -b $file.tx.polyA -x 1 -y 1 > Fl/$file.fl.tx.tsv

        echo '++++++ Aquiring geneIds and txIds ++++++'
        # --------------------------------------------------------------------
        # GTF from mm10.bed12
        # --------------------------------------------------------------------
        zcat $file.bed12.gz | ../../Utils/bed12togff - > $file.lncRNAs.mm10.gtf

        # --------------------------------------------------------------------
        # build the gene ids
        # --------------------------------------------------------------------
        echo $full_name
        bedtools intersect -s -wao -a $file.lncRNAs.mm10.gtf -b $file.lncRNAs.mm10.gtf | ../../Utils/buildLoci.pl - > $file.lncRNAs.mm10.gene.tmp.gtf
        ../../Utils/gff2gff.pl $file.lncRNAs.mm10.gene.tmp.gtf > $file.lncRNAs.mm10.gene.gtf

        # --------------------------------------------------------------------
        # get the gene ids from the mm10.gtfs
        # --------------------------------------------------------------------
        cat $file.lncRNAs.mm10.gene.gtf | awk '{print $10}' | sed 's/;//g' | sed 's/"//g' | sort | uniq > $file.buildLoci.geneIds

        # --------------------------------------------------------------------
        # Get the txs Ids
        # --------------------------------------------------------------------
        cat $file.lncRNAs.mm10.gene.gtf | awk '$3=="exon"{print $12}' | sed 's/;//g'| sed 's/"//g'| sort | uniq > $file.buildLoci.txIds

        echo '>> Done <<'
        ###################################
        echo 'Generating plot input data'
        # --------------------------------------------------------------------
        fl=`cat Fl/$file.fl.tx.tsv | wc -l`
        total=`zcat $file.bed12.gz | cut -f4 | sort | uniq | wc -l`
        propFl=`echo $fl / $total | bc -l`
        gene=`cat $file.buildLoci.geneIds | sort | uniq | wc -l`
        tx=`cat $file.buildLoci.txIds | sort | uniq | wc -l`
        nbTx=`echo $tx / $gene | bc -l`
        echo -e "$file\t$propFl\t$gene\t$nbTx" >> ../figure_s11d_plot_input.tsv

    done


##########################################################################################
# Recount Support
##########################################################################################
# Copy necessary files and transform them inplace
## Recount data
zcat ../../Data/Additional/m38_recount3.bed.gz | awk '{print $1"\t"$2+1"\t"$3"\t"$5"\t"$6}' > Recount_support/m38_recount3.pseudo.bed

# Run intersecting: catalogs with recount data
echo "Running intersection"
# Create results.tsv with a header
echo -e "catalogue\ttotal\tsupported\tunsupported\tsupported_percentage" > ../figure_s11d_recount_input.tsv
for file in `ls ./M_catalogs/*gtf.gz | awk -F'/' '{print $NF}' | sed 's/.gtf.gz//g'`
  do
      echo "Name: $file"
      # Convert to gtf for compatibility
      zcat $file.bed12.gz | ../../Utils/bed12togff - > Recount_support/$file.gtf

      echo "~~~ Preparing introns ~~~"
      # For introns to be compatible with recount data, they need to be modified: start+1 stop-1
      # I included this in modified version on make_introns.awk script
      sort -k12,12 -k4,4n -k5,5n Recount_support/$file.gtf | awk -v fldgn=10 -v fldtr=12 -f ../../Utils/make_introns_for_recount.awk | awk '{print $1"\t"$4"\t"$5"\t"$7"\t"$12}' | sed 's/;//g' | sed 's/"//g' > Recount_support/$file.introns
      echo "+++ Done +++"
      echo "~~~ Right-outer joining recount data with Recount_support/$file introns  ~~~"
      # Right-outer join seems easier for join.py script than left-outer join which is I guess not supported
      # Results are as expected, but be sure that u submit recount as "file a"
      ../../Utils/join.py -a Recount_support/m38_recount3.pseudo.bed -b Recount_support/$file.introns -x 1,2,3,5 -y 1,2,3,4 -u > Recount_support/${file}_with_recount.tsv
      # Calculate supported and unsupported tx
      total_tx=`cat Recount_support/$file.introns | awk '{print $5}' | sort | uniq | wc -l`
      # Fortunately as < 50 awk prints also "-" that were assigned to unmatched introns
      not_supported_tx=`cat Recount_support/${file}_with_recount.tsv | awk '$6 < 50 {print $5}' | sort | uniq | wc -l`
      supported_tx=$((total_tx-not_supported_tx))
      supported_tx_percentage=`echo "scale=2 ; $supported_tx / $total_tx" | bc`
      # Save results to tsv
      echo -e "$file\t$total_tx\t$supported_tx\t$not_supported_tx\t$supported_tx_percentage" >> ../figure_s11d_recount_input.tsv
      echo "+++ Done +++"
  done

# Return
echo '################## COMPLETED ##################'
```

# Load R libraries
```{r}
if (!require("pacman")) install.packages("pacman")
pacman::p_load(
  ggplot2,
  scales,
  dplyr,
  ggrepel,
  scatterpie
)
```

# Rename catalogs here
```{r}
# Clean the env
rm(list = ls())

df <- read.table("figure_s11d_plot_input.tsv", header = F, as.is = T, sep = "\t") %>%
  rename(
    Catalogue = "V1",
    fl = "V2",
    gene = "V3",
    Isoforms = "V4"
  ) %>%
  mutate(Catalogue = ifelse(Catalogue == "noncode", "NONCODE", Catalogue)) %>%
  mutate(Catalogue = ifelse(Catalogue == "refseq", "refSeq", Catalogue)) %>%
  mutate(Catalogue = ifelse(Catalogue == "gencodeM3+", "GENCODEvM3+", Catalogue)) %>%
  mutate(Catalogue = ifelse(Catalogue == "genM16", "GENCODE vM16", Catalogue)) %>%
  mutate(Catalogue = ifelse(Catalogue == "genM36", "GENCODE vM36", Catalogue)) %>%
  mutate(Catalogue = ifelse(Catalogue == "CLS_transcripts", "CLS transcripts", Catalogue)) %>%
  # Add true numbers of tx and genes for Gencode M16 and Gencode M36 (lncRNA)
  # Long non-coding RNA (vM16) - gencode.vM16.long_noncoding_RNAs.gtf.gz
  mutate(gene=ifelse((Catalogue=="GENCODE vM16"), 8715, gene)) %>% 
  mutate(Isoforms=ifelse((Catalogue=="GENCODE vM16"), 13079/8715, Isoforms)) %>%
  # Long non-coding RNA (vM36) gencode.vM36.long_noncoding_RNAs.gtf.gz
  mutate(gene=ifelse((Catalogue=="GENCODE vM36"), 30157, gene)) %>% 
  mutate(Isoforms=ifelse((Catalogue=="GENCODE vM36"), 146802/30157, Isoforms))

# Preview
df
```

# PIE-CHARTS <- Figure S11d
```{r}
cbPalette <- c("#fab22b", "#204f7c","#cccccc",
               "#cccccc", "#cccccc")

# Font colors within boxes
cols = c("GENCODE vM16" = "#FFFFFF",
         "GENCODE vM36" = "#FFFFFF",
         "GENCODEvM3+" = "#000000",
         "NONCODE" = "#000000",
         "refSeq" = "#000000")

df_recount <- read.table("figure_s11d_recount_input.tsv", header=T, as.is=T, sep="\t") %>%
  rename(Catalogue="catalogue") %>%  # to make it compatible with barbie tsv output
  mutate(Catalogue = ifelse(Catalogue=="gencodeM3+","GENCODEvM3+", Catalogue)) %>%
  mutate(Catalogue = ifelse(Catalogue=="noncode","NONCODE", Catalogue)) %>%
  mutate(Catalogue = ifelse(Catalogue=="refseq","refSeq", Catalogue)) %>%
  mutate(Catalogue = ifelse(Catalogue=="genM16","GENCODE vM16", Catalogue)) %>%
  mutate(Catalogue = ifelse(Catalogue=="genM36","GENCODE vM36", Catalogue)) %>%
  mutate(Catalogue = ifelse(Catalogue=="Mouse_CapTrap-CLS_lncRNAs","CLS transcripts", Catalogue))

df %>%
  filter(Catalogue=="GENCODE vM16"
         | Catalogue=="GENCODE vM36"
         | Catalogue=="GENCODEvM3+"
         | Catalogue=="NONCODE"
         | Catalogue=="refSeq") -> df_temp

# Split and adjust formatting to piechart requirements
df_recount1 <- df_recount %>% 
  select(Catalogue, supported) %>% 
  rename(value="supported") %>% 
  mutate(component="supported") %>% 
  inner_join(x=.,
             y=df_temp)

df_recount2 <- df_recount %>% 
  select(Catalogue, unsupported) %>% 
  rename(value="unsupported") %>% 
  mutate(component="unsupported") %>% 
  inner_join(x=.,
             y=df_temp)

# Combine
df_combined <- df_recount1 %>% 
  add_row(df_recount2) %>%
  group_by(Catalogue, gene, fl, Isoforms) %>% 
  mutate(pi_area = sqrt(Isoforms/(pi))*7500) %>% # 5000 multiplier was adjusted manually to fit geompoint size *9000 pi_area = sqrt(Isoforms/(2*pi))*8000
  ungroup() %>%
  group_by(Catalogue, gene, fl, Isoforms, pi_area)
  
df_combined
  
## Create Subplots
df.grobs <- df_combined %>% 
  do(subplots = ggplot(., aes(1, value, fill = component)) +
       geom_col(position = "fill", colour = "white") + #, alpha = 0.75
       coord_polar(theta = "y") + 
       theme_void() +
       scale_fill_manual(values=c("#33b516", "#d80032")) + # remember to adjust also fake geom_col on the plot below!
       guides(fill = F)) %>% 
  mutate(subgrobs = list(annotation_custom(ggplotGrob(subplots), 
                      x = gene-pi_area/4, y = fl-pi_area/4, 
                      xmax = gene+pi_area/4, ymax = fl+pi_area/4))) %>% 
  arrange(factor(Catalogue,
                 levels=c("GENCODE vM16",
                          "GENCODE vM36",
                          "GENCODEvM3+",
                          "NONCODE",
                          "refSeq"))) # declared above, needed to control colors

## Display df containing subplots
df.grobs

# Create plot and add all layers
df.grobs %>%
  {ggplot(data = ., aes(gene, fl)) +
       geom_point(data = df_combined, shape = 21,
                 aes(gene*(-1)+4000, fl, size = Isoforms)) + # coordinates were switched, so they wont be displayed in set X scale limits use it to scale legend with pies; isoforms are scaled to produce bigger circle in the legend
      geom_col(data = df_combined,
               aes(0,0, fill = component), 
               colour = "white") +
      .$subgrobs +
      geom_label_repel(aes(label = Catalogue),
                       fill=cbPalette, # declared outside aes, to not generate legend
                       colour=cols,
                       box.padding   = 0.35,
                       size = 4,
                       point.padding = 0.5,
                       segment.color = 'grey50',
                       show.legend = NA) +  # May be used to remove "a" in the legend (not used here)
      scale_x_continuous(lim=c(0,75000), breaks=seq(0,75000, by=25000)) +
      scale_y_continuous(labels=percent, lim=c(0,0.50), breaks=seq(0,0.50, by=0.25)) +
      scale_fill_manual(values=c("#33b516", "#d80032")) + # remember to adjust also subplots above!
      ylab("% Support") +
      theme_bw(base_size = 24) +
      xlab("Number of loci") +
      theme(axis.text.x = element_text(vjust=0.5),legend.text = element_text(size=12.5)) +
      theme(axis.line.x = element_line(colour = "black"),
            axis.line.y = element_line(colour = "black"),
            panel.grid.major = element_blank(),
            panel.grid.minor = element_blank(),
            panel.border = element_blank(),
            panel.background = element_blank(),
            strip.background = element_rect(colour="black",fill="white")) +
      scale_size_area(max_size = 25) +
      guides(fill = guide_legend(override.aes = list(size=6.5),
                                 title = "recount3 support"),
             size = guide_legend(title = "Isoforms")) # Add explicit size guide)
      } -> plot_figure_s11d
  
  
# Preview plot
plot_figure_s11d
```

# Save plot
```{r}
pdf("figure_s11d.pdf", bg = "white", width = 14, height = 9)
plot_figure_s11d
```
