███████╗██╗ ██████╗ ██╗   ██╗██████╗ ███████╗    ██████╗ ██████╗ 
██╔════╝██║██╔════╝ ██║   ██║██╔══██╗██╔════╝    ╚════██╗██╔══██╗
█████╗  ██║██║  ███╗██║   ██║██████╔╝█████╗       █████╔╝██████╔╝
██╔══╝  ██║██║   ██║██║   ██║██╔══██╗██╔══╝       ╚═══██╗██╔══██╗
██║     ██║╚██████╔╝╚██████╔╝██║  ██║███████╗    ██████╔╝██████╔╝
╚═╝     ╚═╝ ╚═════╝  ╚═════╝ ╚═╝  ╚═╝╚══════╝    ╚═════╝ ╚═════╝ 
                                                                 
# Load R libraries
```{r}
if (!require("pacman")) install.packages("pacman")
pacman::p_load(
  ggplot2,
  scales,
  dplyr,
  ggrepel,
  scatterpie
)
```

# Rename catalogs here
```{r}
# Clean the env
rm(list = ls())

df <- read.table("./Figure_3B_input.tsv", header = F, as.is = T, sep = "\t") %>%
  rename(
    Catalogue = "V1",
    fl = "V2",
    gene = "V3",
    Isoforms = "V4"
  ) %>%
  mutate(Catalogue = ifelse(Catalogue == "bigtrans", "BIGTranscriptome", Catalogue)) %>%
  mutate(Catalogue = ifelse(Catalogue == "fantomCat", "Fantom CAT", Catalogue)) %>%
  mutate(Catalogue = ifelse(Catalogue == "mitrans", "MiTranscriptome", Catalogue)) %>%
  mutate(Catalogue = ifelse(Catalogue == "noncode", "NONCODE", Catalogue)) %>%
  mutate(Catalogue = ifelse(Catalogue == "refseq", "RefSeq", Catalogue)) %>%
  mutate(Catalogue = ifelse(Catalogue == "gencode20+", "GENCODE20+", Catalogue)) %>%
  mutate(Catalogue = ifelse(Catalogue == "gen27", "GENCODE (v27)", Catalogue)) %>%
  mutate(Catalogue = ifelse(Catalogue == "gen47", "GENCODE (v47)", Catalogue)) %>%
  mutate(Catalogue = ifelse(Catalogue == "CLS_transcripts", "CLS transcripts", Catalogue)) %>%
  mutate(Catalogue = ifelse(Catalogue == "chess3_1_2_GRCh38_lncRNA", "lncRNA CHESS 3.1.2", Catalogue)) %>%
  mutate(Catalogue = ifelse(Catalogue == "lncRNA_LncBookv2_0_GRCh38", "lncRNA lncBook 2.0", Catalogue)) %>%
  # Add true numbers of SPLICED tx and genes for Gencode 27 and Gencode 47 (lncRNA)
  # For Gencode annotations do not produce loci via buildLoci.pl script
  # Long non-coding RNA (v47) - gen47.gtf.gz
  mutate(gene = ifelse((Catalogue == "GENCODE (v47)"), 30892, gene)) %>%
  mutate(Isoforms = ifelse((Catalogue == "GENCODE (v47)"), 182987 / 30892, Isoforms)) %>%
  # Long non-coding RNA (v27) gen27.gtf.gz
  mutate(gene = ifelse((Catalogue == "GENCODE (v27)"), 12307, gene)) %>%
  mutate(Isoforms = ifelse((Catalogue == "GENCODE (v27)"), 23997 / 12307, Isoforms))

# Preview
df
```

# PIE-CHARTS <- Figure 3B
```{r}
cbPalette <- c(
  "#fab22b", "#204f7c", "#277961", "#cccccc", "#cccccc",
  "#cccccc", "#cccccc", "#cccccc", "#cccccc",
  "#cccccc", "#cccccc"
)

# Font colors within boxes
cols <- c(
  "GENCODE (v27)" = "#FFFFFF",
  "GENCODE (v47)" = "#FFFFFF",
  "CLS transcripts" = "#FFFFFF",
  "BIGTranscriptome" = "#000000",
  "GENCODE20+" = "#000000",
  "Fantom CAT" = "#000000",
  "MiTranscriptome" = "#000000",
  "NONCODE" = "#000000",
  "RefSeq" = "#000000",
  "lncRNA CHESS 3.1.2" = "#000000",
  "lncRNA lncBook 2.0" = "#000000"
)

df_recount <- read.table("./Figure_3B_recount.tsv", header = T, as.is = T, sep = "\t") %>%
  rename(Catalogue = "catalogue") %>% # to make it compatible with barbie tsv output
  mutate(Catalogue = ifelse(Catalogue == "bigtrans", "BIGTranscriptome", Catalogue)) %>%
  mutate(Catalogue = ifelse(Catalogue == "gencode20+", "GENCODE20+", Catalogue)) %>%
  mutate(Catalogue = ifelse(Catalogue == "fantomCat", "Fantom CAT", Catalogue)) %>%
  mutate(Catalogue = ifelse(Catalogue == "mitrans", "MiTranscriptome", Catalogue)) %>%
  mutate(Catalogue = ifelse(Catalogue == "noncode", "NONCODE", Catalogue)) %>%
  mutate(Catalogue = ifelse(Catalogue == "refseq", "RefSeq", Catalogue)) %>%
  mutate(Catalogue = ifelse(Catalogue == "gen27", "GENCODE (v27)", Catalogue)) %>%
  mutate(Catalogue = ifelse(Catalogue == "gen47", "GENCODE (v47)", Catalogue)) %>%
  mutate(Catalogue = ifelse(Catalogue == "chess3_1_2_GRCh38_lncRNA", "lncRNA CHESS 3.1.2", Catalogue)) %>%
  mutate(Catalogue = ifelse(Catalogue == "lncRNA_LncBookv2_0_GRCh38", "lncRNA lncBook 2.0", Catalogue)) %>%
  mutate(Catalogue = ifelse(Catalogue == "CLS_transcripts", "CLS transcripts", Catalogue))

df %>%
  filter(Catalogue == "GENCODE (v27)" |
    Catalogue == "GENCODE (v47)" |
    Catalogue == "CLS transcripts" |
    Catalogue == "BIGTranscriptome" |
    Catalogue == "GENCODE20+" |
    Catalogue == "Fantom CAT" |
    Catalogue == "MiTranscriptome" |
    Catalogue == "NONCODE" |
    Catalogue == "RefSeq" |
    Catalogue == "lncRNA CHESS 3.1.2" |
    Catalogue == "lncRNA lncBook 2.0") -> df_temp

# Split and adjust formatting to piechart requirements
df_recount1 <- df_recount %>%
  select(Catalogue, supported) %>%
  rename(value = "supported") %>%
  mutate(component = "supported") %>%
  inner_join(
    x = .,
    y = df_temp
  )

df_recount2 <- df_recount %>%
  select(Catalogue, unsupported) %>%
  rename(value = "unsupported") %>%
  mutate(component = "unsupported") %>%
  inner_join(
    x = .,
    y = df_temp
  )

# Combine
df_combined <- df_recount1 %>%
  add_row(df_recount2) %>%
  group_by(Catalogue, gene, fl, Isoforms) %>%
  mutate(pi_area = sqrt(Isoforms / (pi)) * 7500) %>%
  ungroup() %>%
  group_by(Catalogue, gene, fl, Isoforms, pi_area)

df_combined

## Create Subplots
df.grobs <- df_combined %>%
  do(subplots = ggplot(., aes(1, value, fill = component)) +
    geom_col(position = "fill", colour = "white") +
    coord_polar(theta = "y") +
    theme_void() +
    scale_fill_manual(values = c("#33b516", "#d80032")) +
    guides(fill = F)) %>%
  mutate(subgrobs = list(annotation_custom(ggplotGrob(subplots),
    x = gene - pi_area / 4, y = fl - pi_area / 4,
    xmax = gene + pi_area / 4, ymax = fl + pi_area / 4
  ))) %>%
  arrange(factor(Catalogue,
    levels = c(
      "GENCODE (v27)",
      "GENCODE (v47)",
      "CLS transcripts",
      "BIGTranscriptome",
      "GENCODE20+",
      "CLS",
      "Fantom CAT",
      "MiTranscriptome",
      "NONCODE",
      "RefSeq",
      "lncRNA CHESS 3.1.2",
      "lncRNA lncBook 2.0"
    )
  )) # declared above, needed to control colors

## Display df containing subplots
df.grobs

# Create plot and add all layers
df.grobs %>%
  {
    ggplot(data = ., aes(gene, fl)) +
      geom_point(
        data = df_combined, shape = 21,
        aes(gene * (-1) + 4000, fl, size = Isoforms)
      ) +
      geom_col(
        data = df_combined,
        aes(0, 0, fill = component),
        colour = "white"
      ) +
      .$subgrobs +
      geom_label_repel(aes(label = Catalogue),
        fill = cbPalette, # declared outside aes, to not generate legend
        colour = cols,
        box.padding = 0.35,
        size = 4,
        point.padding = 0.5,
        segment.color = "grey50",
        show.legend = NA
      ) +
      scale_x_continuous(lim = c(0, 75000), breaks = seq(0, 75000, by = 25000)) +
      scale_y_continuous(labels = percent, lim = c(0, 0.50), breaks = seq(0, 0.50, by = 0.25)) +
      scale_fill_manual(values = c("#33b516", "#d80032")) +
      ylab("% Support") +
      theme_bw(base_size = 24) +
      xlab("Number of loci") +
      theme(axis.text.x = element_text(vjust = 0.5), legend.text = element_text(size = 12.5)) +
      theme(
        axis.line.x = element_line(colour = "black"),
        axis.line.y = element_line(colour = "black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        strip.background = element_rect(colour = "black", fill = "white")
      ) +
      scale_size_area(max_size = 25) +
      guides(
        fill = guide_legend(
          override.aes = list(size = 6.5),
          title = "Recount support"
        ),
        size = guide_legend(title = "Isoforms")
      )
  } -> Figure_3B_plot


# Preview plot
Figure_3B_plot
```
# Save plot
```{r}
pdf("./Figure_3B_plot.pdf", bg = "white", width = 14, height = 9)
Figure_3B_plot
```