███████╗██╗ ██████╗ ██╗   ██╗██████╗ ███████╗    ███████╗ ██╗███████╗ █████╗      █████╗ ███╗   ██╗██████╗     ███████╗ ██╗███████╗ ██████╗
██╔════╝██║██╔════╝ ██║   ██║██╔══██╗██╔════╝    ██╔════╝███║╚════██║██╔══██╗    ██╔══██╗████╗  ██║██╔══██╗    ██╔════╝███║╚════██║██╔════╝
█████╗  ██║██║  ███╗██║   ██║██████╔╝█████╗      ███████╗╚██║    ██╔╝███████║    ███████║██╔██╗ ██║██║  ██║    ███████╗╚██║    ██╔╝██║     
██╔══╝  ██║██║   ██║██║   ██║██╔══██╗██╔══╝      ╚════██║ ██║   ██╔╝ ██╔══██║    ██╔══██║██║╚██╗██║██║  ██║    ╚════██║ ██║   ██╔╝ ██║     
██║     ██║╚██████╔╝╚██████╔╝██║  ██║███████╗    ███████║ ██║   ██║  ██║  ██║    ██║  ██║██║ ╚████║██████╔╝    ███████║ ██║   ██║  ╚██████╗
╚═╝     ╚═╝ ╚═════╝  ╚═════╝ ╚═╝  ╚═╝╚══════╝    ╚══════╝ ╚═╝   ╚═╝  ╚═╝  ╚═╝    ╚═╝  ╚═╝╚═╝  ╚═══╝╚═════╝     ╚══════╝ ╚═╝   ╚═╝   ╚═════╝
                                                                                                                                           
# Load R libraries
```{r}
if (!require("pacman")) install.packages("pacman")
pacman::p_load(ggplot2,
               scales,
               dplyr,
               tidyr,
               ggrepel)
```

# Prepare environment and run tmerge
```{bash echo=FALSE, engine.opts='-l'}
# Remove old results
rm -r ../Data/.temp/Figure_S17A+S17C/

# Handle errors
set -e          # exit on any non-0 exit status
set -o pipefail # exit on any non-0 exit status in pipe
#set -m

# Create directory tree
mkdir ../Data/.temp/Figure_S17A+S17C/
mkdir ../Data/.temp/Figure_S17A+S17C/H_catalogs/
mkdir ../Data/.temp/Figure_S17A+S17C/H_tmerged/

# Move to working dir
cd ../Data/.temp/Figure_S17A+S17C/

# Copy necessary files
## recount data
zcat ../../Source/Additional/recount3_final.pass1.gff3.gz | awk '{print $1"\t"$4"\t"$5"\t"$6"\t"$7}' > recount3_final.pass1.pseudo.bed
## Full mT
cp ../../Source/masterTable/Hv3_masterTable_refined.gtf.gz .
# Targetdesign
cp ../../Source/masterTable/Hv3_CLS3_targetDesign.gtf.gz .
# Human Catalogs only main chr etc # now we remove dot strand and excluding chrM!!!
zcat ../../Source/Catalogs/bigtrans.gtf.gz | ../../../Utils/gff2bed_full.pl - | awk '$1 ~ /^chr[0-9XY]{1,2}$/ {print $0}' > ./H_catalogs/bigtrans.hg38.bed12
zcat ../../Source/Catalogs/fantomCat.gtf.gz | ../../../Utils/gff2bed_full.pl - | awk '$1 ~ /^chr[0-9XY]{1,2}$/ {print $0}' > ./H_catalogs/fantomCat.hg38.bed12
zcat ../../Source/Catalogs/mitrans.gtf.gz | ../../../Utils/gff2bed_full.pl - | awk '$1 ~ /^chr[0-9XY]{1,2}$/ {print $0}' > ./H_catalogs/mitrans.hg38.bed12
zcat ../../Source/Catalogs/noncode.gtf.gz | ../../../Utils/gff2bed_full.pl - | awk '$1 ~ /^chr[0-9XY]{1,2}$/ {print $0}' > ./H_catalogs/noncode.hg38.bed12
zcat ../../Source/Catalogs/refseq.gtf.gz | ../../../Utils/gff2bed_full.pl - | awk '$1 ~ /^chr[0-9XY]{1,2}$/ {print $0}' > ./H_catalogs/refseq.hg38.bed12
zcat ../../Source/Catalogs/gencode20+.gtf.gz | ../../../Utils/gff2bed_full.pl - | awk '$1 ~ /^chr[0-9XY]{1,2}$/ {print $0}' > ./H_catalogs/gencode20+.hg38.bed12

# Create list of all catalog transcript ids (all, not only spliced and main chr)
zcat ../../Source/Catalogs/bigtrans.gtf.gz ../../Source/Catalogs/fantomCat.gtf.gz ../../Source/Catalogs/mitrans.gtf.gz ../../Source/Catalogs/noncode.gtf.gz ../../Source/Catalogs/refseq.gtf.gz ../../Source/Catalogs/gencode20+.gtf.gz | ../../../Utils/gff2bed_full.pl - | awk '{print $4}' | sort | uniq > temp.H_concat_catalogs.tx.ids

# Transform data
## Concatenate human catalogs into 1 file (ENST00000610119.1 has inconsistent strand); remove chrM
cat ./H_catalogs/bigtrans.hg38.bed12 ./H_catalogs/gencode20+.hg38.bed12 ./H_catalogs/fantomCat.hg38.bed12 ./H_catalogs/mitrans.hg38.bed12 ./H_catalogs/noncode.hg38.bed12 ./H_catalogs/refseq.hg38.bed12 | sort | uniq | grep -vFw "ENST00000610119.1" | awk '$1!="chrM"' | grep "\S" > ./H_catalogs/H_concat_catalogs.hg38.bed12

## Extracting target_ids from targetDesign (take only catalogs, cuz other ids are messed up after stripping - eg. can strip to: "1000")
zcat Hv3_CLS3_targetDesign.gtf.gz | extract.gtf.tags.sh - target_id,target_category | awk '$2=="longNonCoding"{print$1}' | sort | uniq | grep "\S" | gzip > Hv3_targets.ids.gz

## Extract transcript_ids that were targeted from target_ids (remove catalog name from first field and last 4 fields, fix broken mitrans ids ending with .+)
zcat Hv3_targets.ids.gz | awk -F'.' '{ OFS="."; for (i=2; i<=NF-4; i++) printf $i (i<NF-4?OFS:"\n") }' | sed 's/.+//g' | sort | uniq | grep "\S" > Hv3_targeted_catalog_tx.ids

## Extract target_ids detected in mT
### Extract target_ids with artifact tags
zcat Hv3_masterTable_refined.gtf.gz | extract.gtf.tags.sh - target,artifact | grep "\S" | gzip > Hv3_masterTable_refined_targets_with_artifacts.tsv.gz
### Extract genuine target_ids
zcat Hv3_masterTable_refined_targets_with_artifacts.tsv.gz | awk '$2=="no"{print$1}' | sed 's/,/\n/g' | sort | uniq | grep "\S" > Hv3_masterTable_refined_genuine_target.ids
### Extract NOT genuine target_ids (with "recountSlt50" tag)
zcat Hv3_masterTable_refined_targets_with_artifacts.tsv.gz | awk '$2=="no" || $2=="recountSlt50" {print $1}' | sed 's/,/\n/g' | sort | uniq | grep "\S" > Hv3_masterTable_refined_other_target.ids

## Extract transcript_ids from targets
### Genuine
cat Hv3_masterTable_refined_genuine_target.ids | awk -F'.' '{ OFS="."; for (i=2; i<=NF-4; i++) printf $i (i<NF-4?OFS:"\n") }' | sed 's/.+//g' | sort | uniq | grep "\S" >  Hv3_masterTable_refined_genuine_catalog_tx.ids
### Other
cat Hv3_masterTable_refined_other_target.ids | awk -F'.' '{ OFS="."; for (i=2; i<=NF-4; i++) printf $i (i<NF-4?OFS:"\n") }' | sed 's/.+//g' | sort | uniq | grep "\S" >  Hv3_masterTable_refined_other_catalog_tx.ids

## Extract target_ids that belong to catalogs and were detected
### in genuine mT
cat Hv3_targeted_catalog_tx.ids | grep -Fwf Hv3_masterTable_refined_genuine_catalog_tx.ids | sort | uniq | grep "\S" > H_catalog_tx_ids_detected_in_genuinemT.ids
### in notgenuine (other) mT (accepting recountSlt50 artifacts)
cat Hv3_targeted_catalog_tx.ids | grep -Fwf Hv3_masterTable_refined_other_catalog_tx.ids | sort | uniq | grep "\S" > H_catalog_tx_ids_detected_in_othermT.ids

## Extract target_ids that belong to catalogs and were not detected
### in genuine mT
cat Hv3_targeted_catalog_tx.ids | grep -vFwf Hv3_masterTable_refined_genuine_catalog_tx.ids | sort | uniq | grep "\S" > H_catalog_tx_ids_NOTdetected_in_genuinemT.ids
### in notgenuine (other) mT (accepting recountSlt50 artifacts)
cat Hv3_targeted_catalog_tx.ids | grep -vFwf Hv3_masterTable_refined_other_catalog_tx.ids | sort | uniq | grep "\S" > H_catalog_tx_ids_NOTdetected_in_othermT.ids

## Check if all extracted ids may be found in catalogs
set +e
cat Hv3_targeted_catalog_tx.ids | grep -vFwf temp.H_concat_catalogs.tx.ids > Hv3_MISSING_targeted_catalog_tx.ids
missed_tx_ids=`cat Hv3_MISSING_targeted_catalog_tx.ids | wc -l`
if [ $missed_tx_ids -gt 0 ]
then
    echo 'SOME OF THE TARGET_IDS WERE NOT CORRECTLY CONVERTED TO CATALOG TRANSCRIPT_IDS!!'
    echo "Detected $missed_tx_ids wrong entries!"
    exit 1
else
    echo 'Everything is fine. All target_ids were sucessfully converted to catalog transcript_ids.'
fi
#set -e

## Prepare config file
for file in `ls ./H_catalogs/*hg38.bed12`; do echo $file >> ./H_tmerge.config; done


## Tmerge transcripts to get non-redundant intronChains
while read file || [ -n "$file" ]
  do
    echo "~~~~ $file ~~~~"
    name=`echo $file | awk -F"/" '{print $3}' | awk -F"." '{print $1}'`
    echo "---- $name ----"
    echo "Tmerging..."
    cat $file | ../../../Utils/bed12togff - | ../../../Utils/sortgff - | ../../../Utils/tmerge --exonOverhangTolerance 8 --tmPrefix IC - | grep '\S' | gzip > ./H_tmerged/$name.tmerged.gtf.gz
    echo "Preparing tmerged targeted slice..."
    cat $file | grep -Fwf Hv3_targeted_catalog_tx.ids | grep "\S" | ../../../Utils/bed12togff - | ../../../Utils/sortgff - | ../../../Utils/tmerge --exonOverhangTolerance 8 --tmPrefix IC - | grep '\S' | gzip > ./H_tmerged/$name.targeted.slice.tmerged.gtf.gz
    echo "Preparing tmerged detected in other mT slice..." # having "no" and "recountSlt50" tags
    cat $file | grep -Fwf H_catalog_tx_ids_detected_in_othermT.ids | grep "\S" | ../../../Utils/bed12togff - | ../../../Utils/sortgff - | ../../../Utils/tmerge --exonOverhangTolerance 8 --tmPrefix IC - | grep '\S' | gzip > ./H_tmerged/$name.detectedInOther.slice.tmerged.gtf.gz
    echo "Preparing tmerged detected in genuine mT slice..."
    cat $file | grep -Fwf H_catalog_tx_ids_detected_in_genuinemT.ids | grep "\S" | ../../../Utils/bed12togff - | ../../../Utils/sortgff - | ../../../Utils/tmerge --exonOverhangTolerance 8 --tmPrefix IC - | grep '\S' | gzip > ./H_tmerged/$name.detectedInGenuine.slice.tmerged.gtf.gz
    echo "Preparing tmerged NOT detected in other mT slice..." # having "no" and "recountSlt50" tags
    cat $file | grep -Fwf H_catalog_tx_ids_NOTdetected_in_othermT.ids | grep "\S" | ../../../Utils/bed12togff - | ../../../Utils/sortgff - | ../../../Utils/tmerge --exonOverhangTolerance 8 --tmPrefix IC - | grep '\S' | gzip > ./H_tmerged/$name.NOTdetectedInOther.slice.tmerged.gtf.gz
    echo "Preparing tmerged NOT detected in genuine mT slice..."
    cat $file | grep -Fwf H_catalog_tx_ids_NOTdetected_in_genuinemT.ids | grep "\S" | ../../../Utils/bed12togff - | ../../../Utils/sortgff - | ../../../Utils/tmerge --exonOverhangTolerance 8 --tmPrefix IC - | grep '\S' | gzip > ./H_tmerged/$name.NOTdetectedInGenuine.slice.tmerged.gtf.gz
    echo "Adding numbers to the output file..."
    all_IC=`zcat ./H_tmerged/$name.tmerged.gtf.gz | extract.gtf.tags.sh - transcript_id | sort | uniq | grep "\S" | wc -l`
    all_TX=`cat $file | awk '{print $4}' | sort | uniq | grep "\S" | wc -l`
    targeted_IC=`zcat ./H_tmerged/$name.targeted.slice.tmerged.gtf.gz | extract.gtf.tags.sh - transcript_id | sort | uniq | grep "\S" | wc -l`
    targeted_TX=`cat $file | grep -Fwf Hv3_targeted_catalog_tx.ids | awk '{print $4}' | sort | uniq | grep "\S" | wc -l`
    detected_other_IC=`zcat ./H_tmerged/$name.detectedInOther.slice.tmerged.gtf.gz | extract.gtf.tags.sh - transcript_id | sort | uniq | grep "\S" | wc -l`
    detected_other_TX=`cat $file | grep -Fwf H_catalog_tx_ids_detected_in_othermT.ids | awk '{print $4}' | sort | uniq | grep "\S" | wc -l`
    detected_genuine_IC=`zcat ./H_tmerged/$name.detectedInGenuine.slice.tmerged.gtf.gz | extract.gtf.tags.sh - transcript_id | sort | uniq | grep "\S" | wc -l`
    detected_genuine_TX=`cat $file | grep -Fwf H_catalog_tx_ids_detected_in_genuinemT.ids | awk '{print $4}' | sort | uniq | grep "\S" | wc -l`
    NOTdetected_other_IC=`zcat ./H_tmerged/$name.NOTdetectedInOther.slice.tmerged.gtf.gz | extract.gtf.tags.sh - transcript_id | sort | uniq | grep "\S" | wc -l`
    NOTdetected_other_TX=`cat $file | grep -Fwf H_catalog_tx_ids_NOTdetected_in_othermT.ids | awk '{print $4}' | sort | uniq | grep "\S" | wc -l`
    NOTdetected_genuine_IC=`zcat ./H_tmerged/$name.NOTdetectedInGenuine.slice.tmerged.gtf.gz | extract.gtf.tags.sh - transcript_id | sort | uniq | grep "\S" | wc -l`
    NOTdetected_genuine_TX=`cat $file | grep -Fwf H_catalog_tx_ids_NOTdetected_in_genuinemT.ids | awk '{print $4}' | sort | uniq | grep "\S" | wc -l`
    echo -e "$name\t$all_IC\t$all_TX\t$targeted_IC\t$targeted_TX\t$detected_other_IC\t$NOTdetected_other_IC\t$detected_other_TX\t$NOTdetected_other_TX\t$detected_genuine_IC\t$NOTdetected_genuine_IC\t$detected_genuine_TX\t$NOTdetected_genuine_TX" >> H_intronChains_target_detection.tsv
    echo "++ Done ++"
  done < ./H_tmerge.config
```

# Run recount analysis <- previous box is necessary to run this
```{bash echo=FALSE, engine.opts='-l'}
# Remove old results
rm -r ../Data/.temp/Figure_S17A+S17C/Recount/

# Handle errors
set -e          # exit on any non-0 exit status
set -o pipefail # exit on any non-0 exit status in pipe
#set -m

# Create directory tree
mkdir ../Data/.temp/Figure_S17A+S17C/Recount/
mkdir ../Data/.temp/Figure_S17A+S17C/Recount/H_catalogs_gtf/

# Move to working dir
cd ../Data/.temp/Figure_S17A+S17C/Recount/

# Copy necessary files
## recount data from Tamara
zcat ../../../Source/Additional/recount3_final.pass1.gff3.gz | awk '{print $1"\t"$4"\t"$5"\t"$6"\t"$7}' > recount3_final.pass1.pseudo.bed

## Catalogs from above intronChain part and convert to gff + add some processed stuff
for file in `ls ../H_catalogs/*.hg38.bed12`
  do
    name=`echo $file | awk -F"/" '{print $3}' | awk -F"." '{print $1}'`
    # All tx
    cat $file | ../../../../Utils/bed12togff - | gzip > ./H_catalogs_gtf/$name.gtf.gz
    # All ic
    cp ../H_tmerged/$name.tmerged.gtf.gz ./H_catalogs_gtf/${name}_tmerged.gtf.gz
    # Targeted tx
    cat $file | grep -Fwf ../Hv3_targeted_catalog_tx.ids | grep "\S" | ../../../../Utils/bed12togff - | gzip > ./H_catalogs_gtf/${name}_targeted_slice.gtf.gz
    # Targeted ic
    cp ../H_tmerged/$name.targeted.slice.tmerged.gtf.gz ./H_catalogs_gtf/${name}_targeted_slice_tmerged.gtf.gz
    # Detected other tx
    cat $file | grep -Fwf ../H_catalog_tx_ids_detected_in_othermT.ids | grep "\S" | ../../../../Utils/bed12togff - | gzip > ./H_catalogs_gtf/${name}_detectedInOther.slice.gtf.gz
    # Detected other ic
    cp ../H_tmerged/$name.detectedInOther.slice.tmerged.gtf.gz ./H_catalogs_gtf/${name}_detectedInOther_slice_tmerged.gtf.gz
    # Detected genuine tx
    cat $file | grep -Fwf ../H_catalog_tx_ids_detected_in_genuinemT.ids | grep "\S" | ../../../../Utils/bed12togff - | gzip > ./H_catalogs_gtf/${name}_detectedInGenuine_slice.gtf.gz
    # Detected genuine ic
    cp ../H_tmerged/$name.detectedInGenuine.slice.tmerged.gtf.gz ./H_catalogs_gtf/${name}_detectedInGenuine_slice_tmerged.gtf.gz
    # NOT Detected other tx
    cat $file | grep -Fwf ../H_catalog_tx_ids_NOTdetected_in_othermT.ids | grep "\S" | ../../../../Utils/bed12togff - | gzip > ./H_catalogs_gtf/${name}_NOTdetectedInOther_slice.gtf.gz
    # NOT Detected other ic
    cp ../H_tmerged/$name.NOTdetectedInOther.slice.tmerged.gtf.gz ./H_catalogs_gtf/${name}_NOTdetectedInOther_slice_tmerged.gtf.gz
    # NOT Detected genuine tx
    cat $file | grep -Fwf ../H_catalog_tx_ids_NOTdetected_in_genuinemT.ids | grep "\S" | ../../../../Utils/bed12togff - | gzip > ./H_catalogs_gtf/${name}_NOTdetectedInGenuine_slice.gtf.gz
    # NOT Detected genuine ic
    cp ../H_tmerged/$name.NOTdetectedInGenuine.slice.tmerged.gtf.gz ./H_catalogs_gtf/${name}_NOTdetectedInGenuine_slice_tmerged.gtf.gz
  done

## Create new config file
for catalogue in `ls ./H_catalogs_gtf/*gtf.gz`; do echo $catalogue >> H_recount.config; done

while read file || [ -n "$file" ]
  do
    echo "~~~~ $file ~~~~"
    name=`echo $file | awk -F"/" '{print $3}' | awk -F"." '{print $1}'`
    echo "---- $name ----"
    
    echo "~~~ Preparing introns ~~~"
      # For introns to be compatible with recount data, they need to be modified: start+1 stop-1
      # I included this in modified version on make_introns.awk script
      zcat $file | sort -k12,12 -k4,4n -k5,5n - | awk -v fldgn=10 -v fldtr=12 -f ../../../../Utils/make_introns_for_recount.awk | awk '{print $1"\t"$4"\t"$5"\t"$7"\t"$12}' | sed 's/;//g' | sed 's/"//g' > $name.introns
      echo "+++ Done +++"
      echo "~~~ Right-outer joining recount data with $name introns  ~~~"
      # Right-outer join seems easier for join.py script than left-outer join which is I guess not supported
      # Results are as expected, but be sure that u submit recount as "file a"
      ../../../../Utils/join.py -a recount3_final.pass1.pseudo.bed -b $name.introns -x 1,2,3,5 -y 1,2,3,4 -u > ${name}_with_recount.tsv
      # Calculate supported and unsupported tx - percentage is always calculated for tx that have introns
      total_tx=`cat $name.introns | awk '{print $5}' | sort | uniq | wc -l`
      # Fortunately as < 50 awk prints also "-" that were assigned to unmatched introns
      not_supported_tx=`cat ${name}_with_recount.tsv | awk '$6 < 50 {print $5}' | sort | uniq | wc -l`
      supported_tx=$((total_tx-not_supported_tx))
      supported_tx_percentage=`echo "scale=2 ; $supported_tx / $total_tx" | bc`
      # Save results to tsv
      echo -e "$name\t$total_tx\t$supported_tx\t$not_supported_tx\t$supported_tx_percentage" >> recount.results.tsv
      echo "+++ Done +++"    
  done < ./H_recount.config

```

# Run catalog specificity analysis <- previous box is necessary to run this
```{bash echo=FALSE, engine.opts='-l'}
# Remove old results
rm -r ../Data/.temp/Figure_S17A+S17C/Catalog_specificity/

# Handle errors
set -e          # exit on any non-0 exit status
set -o pipefail # exit on any non-0 exit status in pipe
#set -m

# Create directory tree
mkdir ../Data/.temp/Figure_S17A+S17C/Catalog_specificity/


# Move to working dir
cd ../Data/.temp/Figure_S17A+S17C/Catalog_specificity/

# Extract transcirpt_ids from contrains fiield of tmenrge output (has only one tx_id in contains tmerge output) #| grep -v ","
## For IC not detected in genuine
zcat ../H_tmerged/H_concat_catalogs.NOTdetectedInGenuine.slice.tmerged.gtf.gz | extract.gtf.tags.sh - transcript_id,contains | sort | uniq | grep "\S" > H_extracted_transcripts.ids
## For IC detected in genuine
zcat ../H_tmerged/H_concat_catalogs.detectedInGenuine.slice.tmerged.gtf.gz | extract.gtf.tags.sh - transcript_id,contains | sort | uniq | grep "\S" > H_extracted_transcripts_detected.ids
## For IC from merged-lncRNAs
zcat ../H_tmerged/H_concat_catalogs.tmerged.gtf.gz | extract.gtf.tags.sh - transcript_id,contains | sort | uniq | grep "\S" > H_extracted_transcripts_mergedRNAs.ids


# Create table with tx_ids and catalog name
for file in ../H_catalogs/*bed12
  do
     name=`echo $file | awk -F'/' '{print $NF}' | awk -F '.' '{print $1}'`
     # do not include concatenated catalogs
     if [[ "$name" == "H_concat_catalogs" ]]; then
     continue
     fi
     cat $file | awk -v catalog=$name '{print $4 "\t" catalog}' >> H_transcripts_with_catalog_names.tsv
  done
```

# Load and analyze data <- FIGURE S17A
```{r}
# Clean the env
rm(list = ls())

# Load data
## cropped Hv3_masterTable_refined
df_H_intronChains_target_detection <- read.table("../Data/.temp/Figure_S17A+S17C/H_intronChains_target_detection.tsv", header=F, sep="\t") %>%
  rename(H_catalog_name="V1",
         H_all_ic="V2",
         H_all_tx="V3",
         H_targeted_ic="V4",
         H_targeted_tx="V5",
         H_detected_other_ic="V6",
         H_NOTdetected_other_ic="V7",
         H_detected_other_tx="V8",
         H_NOTdetected_other_tx="V9",
         H_detected_genuine_ic="V10",
         H_NOTdetected_genuine_ic="V11",
         H_detected_genuine_tx="V12",
         H_NOTdetected_genuine_tx="V13")

# Pivot the data for easier plotting
df_H_intronChains_target_detection %>% 
  select(H_catalog_name, H_all_ic, H_targeted_ic, H_detected_other_ic, H_detected_genuine_ic, H_NOTdetected_genuine_ic, H_NOTdetected_other_ic) %>% 
  mutate(H_catalog_name=sub("H_concat_catalogs*", "mergedCatalogs", H_catalog_name)) %>%
  rename(all="H_all_ic", targeted="H_targeted_ic", detected_other="H_detected_other_ic", detected_genuine="H_detected_genuine_ic", NOTdetected_genuine="H_NOTdetected_genuine_ic", NOTdetected_other="H_NOTdetected_other_ic") %>% 
  mutate(H_type="intronChains") %>% 
  tidyr::pivot_longer(
     cols = c("all", "targeted", "detected_other", "detected_genuine", "NOTdetected_genuine", "NOTdetected_other"),
     names_to = "H_status",
     values_to = "H_count") -> df_H_intronChains_target_detection_IC
  
# Pivot the data for easier plotting
df_H_intronChains_target_detection %>% 
  select(H_catalog_name, H_all_tx, H_targeted_tx, H_detected_other_tx, H_detected_genuine_tx, H_NOTdetected_genuine_tx, H_NOTdetected_other_tx) %>% 
  mutate(H_catalog_name=sub("H_concat_catalogs*", "mergedCatalogs", H_catalog_name)) %>%
  rename(all="H_all_tx", targeted="H_targeted_tx", detected_other="H_detected_other_tx", detected_genuine="H_detected_genuine_tx", NOTdetected_genuine="H_NOTdetected_genuine_tx", NOTdetected_other="H_NOTdetected_other_tx") %>% 
  mutate(H_type="transcripts") %>% 
  tidyr::pivot_longer(
     cols = c("all", "targeted", "detected_other", "detected_genuine", "NOTdetected_genuine", "NOTdetected_other"),
     names_to = "H_status",
     values_to = "H_count") -> df_H_intronChains_target_detection_TX

# Preview
df_H_intronChains_target_detection
df_H_intronChains_target_detection_IC
df_H_intronChains_target_detection_TX

#--------------------------------

# Read recount data
df_recount_support <- read.table("../Data/.temp/Figure_S17A+S17C/Recount/recount.results.tsv", header=F, sep="\t") %>%
  rename(H_catalog_name="V1",
         total_tx="V2",
         supported_tx="V3",
         not_supported_tx="V4",
         supported_tx_percentage="V5") %>% 
  rowwise() %>%
  mutate(H_catalog_name=sub("H_concat_catalogs*", "mergedCatalogs", H_catalog_name)) %>%
  mutate(H_status=sub("^[^_]*_", "", H_catalog_name)) %>% 
  mutate(H_type=ifelse((H_status=="detectedInGenuine_slice_tmerged" |
                         H_status=="detectedInOther_slice_tmerged" |
                         H_status=="NOTdetectedInGenuine_slice_tmerged" |
                         H_status=="NOTdetectedInOther_slice_tmerged" |
                         H_status=="targeted_slice_tmerged" |
                         H_status=="tmerged"), "intronChains", "transcripts")) %>% 
  mutate(H_status=ifelse((H_status==H_catalog_name | H_status=="tmerged"), "all", H_status)) %>% 
  mutate(H_status=ifelse((H_status=="detectedInGenuine_slice" | H_status=="detectedInGenuine_slice_tmerged"), "detected_genuine", H_status)) %>%
  mutate(H_status=ifelse((H_status=="detectedInOther" | H_status=="detectedInOther_slice_tmerged"), "detected_other", H_status)) %>%
  mutate(H_status=ifelse((H_status=="NOTdetectedInGenuine_slice" | H_status=="NOTdetectedInGenuine_slice_tmerged"), "NOTdetected_genuine", H_status)) %>%
  mutate(H_status=ifelse((H_status=="NOTdetectedInOther_slice" | H_status=="NOTdetectedInOther_slice_tmerged"), "NOTdetected_other", H_status)) %>%
  mutate(H_status=ifelse((H_status=="targeted_slice" | H_status=="targeted_slice_tmerged"), "targeted", H_status)) %>%
  mutate(H_catalog_name=sub("_.*", "", H_catalog_name)) %>% 
  select(H_catalog_name, supported_tx_percentage, H_status, H_type) %>% 
  rename(recount_support="supported_tx_percentage")

# Preview
df_recount_support


# Plot
df_H_intronChains_target_detection_IC %>%
  inner_join(x=.,
             y=df_recount_support,
             by=c("H_catalog_name", "H_status", "H_type")) %>% 
  # Filter out unnecessary cols
  filter(H_status != "detected_other" & H_status !="NOTdetected_other") %>% 
  rowwise() %>% 
  mutate(H_status=ifelse((H_status=="detected_genuine"), "GENCODE 47", H_status)) %>% 
  mutate(H_status=ifelse((H_status=="NOTdetected_genuine"), "NOT in GENCODE 47", H_status)) %>% 
  mutate(H_catalog_name=ifelse(H_catalog_name=="mergedCatalogs", "Unique transcripts", H_catalog_name)) %>% 
  mutate(H_catalog_name=ifelse(H_catalog_name=="gencode20+", "GENCODE20+", H_catalog_name)) %>%
  
  mutate(H_catalog_name=ifelse(H_catalog_name=="mitrans", "MiTranscriptome", H_catalog_name)) %>%
  mutate(H_catalog_name=ifelse(H_catalog_name=="noncode", "NONCODE", H_catalog_name)) %>%
  mutate(H_catalog_name=ifelse(H_catalog_name=="fantomCat", "FANTOM CAT", H_catalog_name)) %>%
  mutate(H_catalog_name=ifelse(H_catalog_name=="refseq", "RefSeq", H_catalog_name)) %>%
  mutate(H_catalog_name=ifelse(H_catalog_name=="bigtrans", "BIGTranscriptome", H_catalog_name)) %>%
  
  mutate(H_group=ifelse((H_catalog_name=="Unique transcripts"), "All", "Individual lncRNA catalogs")) %>% 
  ggplot(aes(x=reorder(H_catalog_name, -H_count),
             y=H_count,
             fill=reorder(H_status, -H_count))) +
  geom_bar(stat="identity", position=position_dodge()) +
  geom_text(aes(label = paste0(scales::comma(H_count), "\n", "(", scales::percent(recount_support), ")")),
          position = position_dodge(width=0.9),
          vjust=-0.25,
          size = 2.0) +
  theme_bw() +
  ylab('# intron chains') +
  ggtitle(label = "Figure S17A") +
  scale_fill_brewer(palette="Purples", direction = -1) +
  scale_y_continuous(labels = comma, limits = c(0, 600000)) +
  theme(axis.title.x=element_blank(),
        axis.text = element_text(colour = "black", size = 11),
        axis.ticks = element_line(colour = "black"),
        panel.background = element_rect(fill = "white", colour = "grey50"),
        legend.title = element_blank(),
        legend.position = c(0.90, 0.80),
        panel.grid = element_blank(),
        plot.title = element_text(hjust = 0.5, size=11)) +
  facet_grid(cols=vars(H_group), scales = "free_x", space = "free") -> plot_Figure_S17A

# Preview plot
plot_Figure_S17A
```

# Save as pdf
```{r}
pdf(file="./plot_Figure_S17A.pdf", width = 10, height = 4)
plot_Figure_S17A
```


# Join tables and analyze data <- catalog specificity << FIGURE S17C
```{r}
# Clean the env
rm(list = ls())

# Load data
## Transcript_ids not detected in genuine (NOT in GENCODE 47)
df_H_extracted_transcripts <- read.table("../Data/.temp/Figure_S17A+S17C/Catalog_specificity/H_extracted_transcripts.ids",
                                       header=F, sep="\t") %>%
  rename(H_intronChain_id="V1",
         H_transcript_id="V2")

## Preview
df_H_extracted_transcripts

## Transcript_ids detected in genuine (in GENCODE 47)
df_H_extracted_transcripts_detected <- read.table("../Data/.temp/Figure_S17A+S17C/Catalog_specificity/H_extracted_transcripts_detected.ids",
                                       header=F, sep="\t") %>%
  rename(H_intronChain_id="V1",
         H_transcript_id="V2")

## Preview
df_H_extracted_transcripts_detected


#########################################
## Transcript_ids with catalog names
df_H_transcripts_with_catalogs <- read.table("../Data/.temp/Figure_S17A+S17C/Catalog_specificity/H_transcripts_with_catalog_names.tsv",
                                       header=F, sep="\t") %>%
  rename(H_transcript_id="V1",
         H_catalog="V2") %>% 
  mutate(H_catalog = ifelse(H_catalog=="bigtrans","BIGTranscriptome", H_catalog)) %>%
  mutate(H_catalog = ifelse(H_catalog=="gencode20+","GENCODE20+", H_catalog)) %>%
  mutate(H_catalog = ifelse(H_catalog=="fantomCat","FANTOM CAT", H_catalog)) %>%
  mutate(H_catalog = ifelse(H_catalog=="mitrans","MiTranscriptome", H_catalog)) %>%
  mutate(H_catalog = ifelse(H_catalog=="noncode","NONCODE", H_catalog)) %>%
  mutate(H_catalog = ifelse(H_catalog=="refseq","RefSeq", H_catalog))

## Preview
df_H_transcripts_with_catalogs

## Check catalogs
df_H_transcripts_with_catalogs %>% 
  distinct(H_catalog)

#########################################
# Join tables
df_H_joined_tables <- df_H_extracted_transcripts %>% 
  left_join(x=.,
            y=df_H_transcripts_with_catalogs)

# Preview
df_H_joined_tables
##################
# Join tables
df_H_joined_tables_detected <- df_H_extracted_transcripts_detected %>% 
  left_join(x=.,
            y=df_H_transcripts_with_catalogs)

# Preview
df_H_joined_tables_detected


#########################################
# Plot this - catalog composition


cbPalette <- c("lightgrey", "darkgrey","#3E506C","#50868d","#58508d","#bc5090","#ff6362","#ffa602")


df_specific_IC <- df_H_joined_tables %>%
  filter(!is.na(H_catalog)) %>%  # to filter out IC_ids that were merged
  summarise(.by="H_intronChain_id", count = n()) %>% 
  filter(count == 1) %>% 
  select(H_intronChain_id) %>% # to filter out those not merged but still present in more than one set (there are 14 ids like this from fantomCat and Gencode+)
  rbind(df_H_joined_tables %>% # here we filter those that were merged but consist of multiple transcripts from the same catalog
          mutate(category=ifelse((is.na(H_catalog)), "non specific", "specific")) %>% 
          filter(category=="non specific") %>% 
          select(H_intronChain_id, H_transcript_id) %>% 
          separate_rows(H_transcript_id, sep = ",") %>%
          left_join(x=.,
                    y=df_H_transcripts_with_catalogs,
                    by="H_transcript_id") %>% 
          summarise(.by=H_intronChain_id,
                    unique_count=length(unique(H_catalog))) %>% 
          filter(unique_count==1) %>% 
          select(H_intronChain_id)) %>% 
  unique()


df_specific_IC_detected <- df_H_joined_tables_detected %>%
  filter(!is.na(H_catalog)) %>%  # to filter out IC_ids that were merged
  summarise(.by="H_intronChain_id", count = n()) %>% 
  filter(count == 1) %>% 
  select(H_intronChain_id) %>% # to filter out those not merged but still present in more than one set (there are 14 ids like this from fantomCat and Gencode+)
  rbind(df_H_joined_tables %>% # here we filter those that were merged but consist of multiple transcripts from the same catalog
          mutate(category=ifelse((is.na(H_catalog)), "non specific", "specific")) %>% 
          filter(category=="non specific") %>% 
          select(H_intronChain_id, H_transcript_id) %>% 
          separate_rows(H_transcript_id, sep = ",") %>%
          left_join(x=.,
                    y=df_H_transcripts_with_catalogs,
                    by="H_transcript_id") %>% 
          summarise(.by=H_intronChain_id,
                    unique_count=length(unique(H_catalog))) %>% 
          filter(unique_count==1) %>% 
          select(H_intronChain_id)) %>% 
  unique()


#########################################
# Plot this - catalog specificity joined with composition

df_H_joined_tables %>% 
  mutate(category=ifelse((H_intronChain_id %in% df_specific_IC$H_intronChain_id), "specific", "non specific")) %>%  
  select(H_intronChain_id, category) %>% 
  unique() %>% 
  group_by(category) %>% 
  summarise(.groups = "keep", count=n()) %>% 
  ungroup() %>% 
  mutate(percentage=count/sum(count)) %>%
  mutate(H_catalog=ifelse((category=="specific"), "Catalog-specific", "Shared across catalogs")) %>% 
  mutate(category="NOT in GENCODE 47") %>% 
  rbind(df_H_joined_tables %>%
          mutate(category=ifelse((H_intronChain_id %in% df_specific_IC$H_intronChain_id), "specific", "non specific")) %>%  
          select(H_intronChain_id, H_transcript_id, category) %>% 
          separate_rows(H_transcript_id, sep = ",") %>%
          left_join(x=.,
                    y=df_H_transcripts_with_catalogs,
                    by="H_transcript_id") %>%
          select(H_intronChain_id, H_catalog, category) %>% 
          unique() %>% 
          group_by(H_catalog, category) %>% 
          summarise(.groups = "keep", count=n()) %>% 
          group_by(category) %>%
          mutate(percentage=count/sum(count)) %>%
          filter(category=="specific") %>%           # comment out to see composition of non specific
          mutate(category="catalog composition")) %>% 
  
  rbind(df_H_joined_tables_detected %>% 
          mutate(category=ifelse((H_intronChain_id %in% df_specific_IC_detected$H_intronChain_id), "specific", "non specific")) %>%  
          select(H_intronChain_id, category) %>% 
          unique() %>% 
          group_by(category) %>% 
          summarise(.groups = "keep", count=n()) %>% 
          ungroup() %>% 
          mutate(percentage=count/sum(count)) %>%
          mutate(H_catalog=ifelse((category=="specific"), "Catalog-specific", "Shared across catalogs")) %>% 
          mutate(category="GENCODE 47")) %>% 
  ## Plot
  ggplot(aes(x=factor(category, levels=c("GENCODE 47",
                                         "NOT in GENCODE 47",
                                         "catalog composition")), y=percentage, fill=factor(H_catalog, levels=c("Catalog-specific", "Shared across catalogs",
                                                                                                                "BIGTranscriptome", "FANTOM CAT", "MiTranscriptome",
                                                                                                                "RefSeq", "GENCODE20+", "NONCODE")))) +
  geom_bar(stat = "identity") +
  geom_text(aes(label = scales::comma(count)),
            position = position_stack(vjust = 0.5),
            size = 3.5) +
  scale_fill_manual(values = cbPalette) +
  theme_bw() +
  #scale_fill_manual(values=c("#979dac", "#e29578")) +
  ggtitle("Figure S17C") +
  ylab("percentage") +
  scale_y_continuous(labels=percent, lim=c(0,1)) +
  theme(legend.position="top") +
  theme(axis.text.x = element_text(size = 12, colour = "black"), #, angle = 60, hjust = 1,
        axis.ticks.x = element_blank(),
        axis.text.y = element_text(size = 12,  colour = "black"),
        legend.text = element_text(size = 12),
        plot.title = element_text(size = 15, hjust = 0.5),
        legend.title = element_blank(),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 14),
        strip.text.x = element_text(size = 18),
        strip.text.y = element_text(size = 18)) -> plot_Figure_S17C


# Preview plot
plot_Figure_S17C
```

# Save plot - catalog specificity with composition
```{r}
pdf(file="./plot_Figure_S17C.pdf", width = 6, height = 6)
plot_Figure_S17C
```







